lib/active_storage/blob_io.rb

```diff
@@ -0,0 +1,54 @@
+module FormatParser
+  module ActiveStorage
+    class BlobIO
+      def initialize(blob)
+        @blob = blob
+        @service = blob.service
+        @pos = 0
+        @remote_size = false
+      end
+
+      # Emulates IO#seek
+      def seek(offset)
+        @pos = offset
+        0 # always return 0
+      end
+
+      # Emulates IO#pos
+      def pos
+        @pos
+      end
+
+      # Emulates IO#size.
+      def size
+        raise 'Remote size not yet obtained, need to perform at least one read() to retrieve it' unless @remote_size
+        @remote_size
+      end
+
+      def read(n_bytes)
+        range_start = @pos
+        range_end =  @pos + n_bytes - 1
+        http_range = (range_start..range_end)
+
+        # ActiveStorageAdapter#download_chunk method return '' when using range (start > end)
+        # Raise invalid range exception if this is the case to recognise the real empty response
+        raise ArgumentError, "Invalid range from #{range_start} to #{range_end}" if range_start > range_end
+
+        begin
+          body = @service.download_chunk(@blob.key, http_range)
+        rescue Errno::ENOENT
+          raise ArgumentError, "Key #{@blob.key} does not exist"
+        rescue Errno::EINVAL
+          # using negative range for example
+          raise ArgumentError, "Invalid range from #{range_start} to #{range_end}"
+        end
+
+        if body
+          @remote_size = body.bytesize
```

Comment:
This is very likely not to work. If you do a ranged HTTP request, the returned result will probably indeed be a String and it's `bytesize` will tell you the size of the segment you managed to retrieve. This is also the value that will be in the `Content-Length` response header. It is not the size of the actual resource however ⚠️ - the size of the resource is in the `Content-Range` header returned by the server. The header has the shape of `Content-Range: bytes 0-1023/146515` where the last block of digits is the actual "remote size" you are after.

Reply:
I got your point, As I saw that this is the case already within `FormatParser` .. will check how to obtain this header from ActiveStorage response (regardless storage service being used) :+1: